reframe/core/schedulers/flux.py

```diff
@@ -0,0 +1,187 @@
+        # We don't need to submit with a file, so we don't need a preamble.
+        return []
+
+    def make_job(self, *args, **kwargs):
+        return _FluxJob(*args, **kwargs)
+
+    def submit(self, job):
+        """
+        Submit a job to the flux executor.
+        """
+        flux_future = self._fexecutor.submit(job.fluxjob)
+        job._jobid = str(flux_future.jobid())
+        job._submit_time = time.time()
+        job._flux_future = flux_future
+
+    def cancel(self, job):
+        """
+        Cancel a running Flux job.
+        """
+        # Job future cannot cancel once running or completed
+        if not job._flux_future.cancel():
+            # This will raise JobException with event=cancel (on poll)
+            flux.job.cancel(flux.Flux(), job._flux_future.jobid())
+
+        # In testing, cancel is too "soft" - or often the pid is still remaining
+        # after. We thus do a cancel in addition to a kill.
+        try:
+            flux.job.kill(flux.Flux(), job._flux_future.jobid())
+
+        # Job is inactive
+        except EnvironmentError:
+            pass
+        job._is_cancelling = True
+
+    def poll(self, *jobs):
+        """
+        Poll running Flux jobs for updated states.
+        """
+        if jobs:
+            # filter out non-jobs
+            jobs = [job for job in jobs if job is not None]
+
+        if not jobs:
+            return
+
+        # Loop through active jobs and act on status
+        for job in jobs:
+            if job._flux_future.done():
+                # The exit code can help us determine if the job was successful
+                try:
+                    exit_code = job._flux_future.result(0)
+                except flux.job.JobException:
+                    # Currently the only state we see is cancelled here
+                    self.log(f"Job {job.jobid} was likely cancelled.")
+                    job._state = "CANCELLED"
+                    job._cancelled = True
+                except RuntimeError:
+                    # Assume some runtime issue (suspended)
+                    self.log(f"Job {job.jobid} was likely suspended.")
+                    job._state = "SUSPENDED"
+                else:
+                    # the job finished (but possibly with nonzero exit code)
+                    if exit_code != 0:
+                        self.log(f"Job {job.jobid} did not finish successfully")
+
+                    job._state = "COMPLETED"
+
+                job._completed = True
+            elif job.state in waiting_states and job.max_pending_time:
+                if time.time() - job.submit_time >= job.max_pending_time:
+                    self.cancel(job)
+                    job._exception = JobError(
+                        "maximum pending time exceeded", job.jobid
+                    )
+            else:
+                # Otherwise, we are still running
+                job._state = "RUNNING"
+
+    def allnodes(self):
```

Comment:
Nice that you implemented this one, but alone it is not very useful. It is used in conjunction with the `filternodes()` below to support flexible node allocation in tests as well as the `--distribute` option. Essentially, the `JobScheduler.guess_num_tasks` tries to guess the available nodes to run the test on, based on various constraints (e.g., the `access` partition attribute, the `-J` scheduler specific options etc.). You can leave the implementation of `filternodes()` for future work in another PR 😉

Reply:
okay I'll raise NotImplementedError for both.